service/controller/v13/templates/cloudconfig/decrypt_keys_assets_script.go

```diff
@@ -1,7 +1,95 @@
 package cloudconfig
 
 const DecryptKeysAssetsScript = `#!/bin/bash -e
+{{ if eq .EncrypterType "vault" }}
+token_path=/var/token
+nonce_path=/var/nonce
 
+# if curl pkcs7 fails, exit with error logged
+token_exists () {
+  if [ -f $token_path ]; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+token_is_valid() {
+  #  https://www.vaultproject.io/api/auth/token/index.html#lookup-a-token-self-
+  echo "Checking token validity"
+  token_lookup=$(curl \
+    --request GET \
+    --header "X-Vault-Token: $(cat $token_path)" \
+    --write-out %{http_code} \
+    --output /dev/null \
+    {{ .VaultAddress }}/v1/auth/token/lookup-self)
+  if [ "$token_lookup" == "200" ]; then
+      echo "$0 - Valid token found, exiting"
+      return 0
+  else
+      echo "$0 - Invalid token found"
+      return 1
+  fi
+}
+
+main () {
+    if ! token_exists; then
+        aws_login ""
+    elif token_exists && ! token_is_valid; then
+        aws_login "$(cat $nonce_path)"
+    else
+        logger $0 "current vault token is still valid"
+    fi
+
+    echo decrypting keys assets
+    shopt -s nullglob
+    for encKey in $(find /etc/kubernetes/encryption -name "*.enc"); do
+      echo decrypting $encKey
+      f=$(mktemp $encKey.XXXXXXXX)
+      curl \
+        --header "X-Vault-Token: $(cat $token_path)" \
+        --request POST \
+        --data '{"ciphertext": "$(cat $encKey)"}' \
+        {{ .VaultAddress }}/v1/transit/decrypt/{{ .EncryptionKey }} | \
+        jq -r .data.plaintext | base64 -d > $f
+      mv -f $f ${encKey%.enc}
+    done;
+    echo done.
+}
+
+aws_login () {
+    pkcs7=$(curl -s http://169.254.169.254/latest/dynamic/instance-identity/pkcs7 | tr -d '\n')
```

Comment:
This is the EC2 metadata endpoint? For me, a comment would help for readability.

Reply:
Sure, done.